pkg/server/plugin/upstreamauthority/wrapper.go

```diff
@@ -0,0 +1,80 @@
+package upstreamauthority
+
+import (
+	"context"
+	"crypto/x509"
+
+	"github.com/spiffe/spire/pkg/server/plugin/upstreamca"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type wrapper struct {
+	upstreamCA upstreamca.UpstreamCA
+}
+
+// Wrap creates a wrapper for UpstreamCA interface that satisfies the UpstreamAuthority interface
+func Wrap(upstreamCA upstreamca.UpstreamCA) UpstreamAuthority {
+	return &wrapper{upstreamCA: upstreamCA}
+}
+
+// MintX509CA pass MinMintX509CA request to UpstreamCA SubmitCSR RPC, parse SubmitCSRResponse into MintX509CAResponse
+func (w *wrapper) MintX509CA(ctx context.Context, request *MintX509CARequest) (*MintX509CAResponse, error) {
+	// Create a SubmitCSRRequest from MintX509CARequest
+	req := &upstreamca.SubmitCSRRequest{
+		Csr:          request.Csr,
+		PreferredTtl: request.PreferredTtl,
+	}
+
+	// Call upstreamCA SubmitCSR
+	resp, err := w.upstreamCA.SubmitCSR(ctx, req)
+	if err != nil {
+		return nil, makeError(codes.Internal, "unable to submit csr: %v", err)
+	}
+
+	// Creates an array of []byte from response CertChain
+	caChain, err := parseCertificates(resp.SignedCertificate.CertChain)
+	if err != nil {
+		return nil, makeError(codes.Internal, "unable to parse cert chain: %v", err)
+	}
+
+	// Creates an array of []byte from response Bundle
+	roots, err := parseCertificates(resp.SignedCertificate.Bundle)
+	if err != nil {
+		return nil, makeError(codes.Internal, "unable to parse bundle: %v", err)
+	}
+
+	return &MintX509CAResponse{
+		X509CaChain:       caChain,
+		UpstreamX509Roots: roots,
+	}, nil
+}
+
+// PublishJWTKey it is not implemented for wrapper
+func (w *wrapper) PublishJWTKey(ctx context.Context, request *PublishJWTKeyRequest) (*PublishJWTKeyResponse, error) {
+	return nil, makeError(codes.Unimplemented, "publishing upstream is unsupported")
+}
+
+// PublishX509CA it is not implemented for wrapper
+func (w *wrapper) PublishX509CA(ctx context.Context, request *PublishX509CARequest) (*PublishX509CAResponse, error) {
+	return nil, makeError(codes.Unimplemented, "publishing upstream is unsupported")
+}
+
+// parseCertificates parse certificates and return an array with each certificate raw
+func parseCertificates(rawCerts []byte) ([][]byte, error) {
+	var certificates [][]byte
```

Comment:
nit: declare this close to point of use (i.e. just before ranging over certChain)

Reply:
done